client/extensions/woocommerce/app/store-stats/store-stats-referrer-widget/index.js

```diff
@@ -64,25 +74,68 @@ class StoreStatsReferrerWidget extends Component {
 			);
 			return [ primary, <p key="link">{ secondary }</p> ];
 		}
-		return this.isPreCollection( selectedData )
+		return this.isPreCollection( date )
 			? [ translate( 'Referral data isn’t available before Jetpack v5.9 (March 2018)' ) ]
 			: [ translate( 'No referral activity on this date' ) ];
 	}
 
+	paginate = data => {
+		const { paginate, limit } = this.props;
+		if ( ! paginate ) {
+			return data.slice( 0, limit || data.length );
+		}
+		const { page } = this.state;
+		const start = ( page - 1 ) * limit;
+		const end = start + limit;
+		return data.slice( start, end );
+	};
+
+	onPageClick = pageNumber => {
+		this.setState( {
+			page: pageNumber,
+		} );
+	};
+
+	setPage( { selectedIndex, limit } ) {
+		if ( this.props.paginate ) {
+			this.setState( {
+				page: Math.floor( selectedIndex / limit ) + 1,
+			} );
+		}
+	}
+
+	componentWillMount() {
+		this.setPage( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		this.setPage( nextProps );
+	}
+
 	render() {
-		const { data, selectedDate, translate, unit, slug, queryParams } = this.props;
+		const {
+			data,
+			selectedDate,
+			translate,
+			unit,
+			slug,
+			queryParams,
+			limit,
+			paginate,
+			selectedReferrer,
+		} = this.props;
+		const { page } = this.state;
 		const basePath = '/store/stats/referrers';
-		const selectedData = find( data, d => d.date === selectedDate ) || { data: [] };
-		if ( selectedData.data.length === 0 ) {
-			const messages = this.getEmptyDataMessage( selectedData );
+		if ( data.length === 0 ) {
+			const messages = this.getEmptyDataMessage( selectedDate );
 			return (
 				<Card className="store-stats-referrer-widget stats-module is-showing-error has-no-data">
 					<ErrorPanel message={ messages.shift() }>{ messages }</ErrorPanel>
 				</Card>
 			);
 		}
-		const sortedAndTrimmedData = sortBySales( selectedData.data, 5 );
-		const extent = [ 0, d3Max( sortedAndTrimmedData.map( d => d.sales ) ) ];
+		const paginatedData = this.paginate( data );
+		const extent = [ 0, d3Max( paginatedData.map( d => d.sales ) ) ];
```

Comment:
I wonder if we shouldn't be using the extent as it relates to `data` rather than `paginatedData`? As you page through it may feel as though a referrer on a lower page is comparable to one on a higher page (visually).

Reply:
This would include persisting the extent of ALL the data when searching/filtering the data.